Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift

```diff
@@ -0,0 +1,184 @@
+
+            self.swizzle(self.method) { typedCurrentImp -> BlockIMP in
+                return { impSelf, impURLRequest, impCompletion -> URLSessionDataTask in
+                    guard let interceptionResult = interceptor(impURLRequest.url) else {
+                        return typedCurrentImp(impSelf, Self.selector, impURLRequest, impCompletion)
+                    }
+
+                    weak var blockTask: URLSessionDataTask? = nil
+                    let modifiedCompletion: CompletionHandler = { origData, origResponse, origError in
+                        impCompletion(origData, origResponse, origError)
+                        blockTask?.payload?(.completed)
+                    }
+                    let modifiedRequest = impURLRequest.merging(interceptionResult.httpHeaders)
+                    let task = typedCurrentImp(impSelf, Self.selector, modifiedRequest, modifiedCompletion)
+                    try? resumeSwizzler.swizzleIfNeeded(in: task)
+                    task.payload = interceptionResult.taskObserver
+                    blockTask = task
+                    return task
+                }
+            }
+        }
+    }
+
+    class Resume: MethodSwizzler <
+        @convention(c) (URLSessionTask, Selector) -> Void,
+        @convention(block) (URLSessionTask) -> Void
+    > {
+        private static let selector = #selector(URLSessionTask.resume)
+
+        /// NOTE: RUMM-452
+        /// URLSessionTask.resume is not called by its subclasses!
+        /// Therefore, we swizzle this method in the subclass.
+        /// This is unlike swizzling dataTaskURL/Request: in URLSession base class
+        func swizzleIfNeeded(in task: URLSessionTask) throws {
+            guard let taskClass = object_getClass(task) else {
+                userLogger.error("Unable to swizzle `URLSessionTask`: \(task) - the trace will not be created.")
+                return
+            }
+            let foundMethod = try Self.findMethod(with: Self.selector, in: taskClass)
+            // NOTE: RUMM-452 We will probably not need to swizzle tasks every time
+            // "onlyIfNonSwizzled: true" lets us perform swizzling in given class if not done before
+            typealias BlockIMP = @convention(block) (URLSessionTask) -> Void
+            swizzle(
+                foundMethod,
+                impProvider: { currentTypedImp -> BlockIMP in
+                    return { impSelf in
+                        impSelf.payload?(.starting)
+                        return currentTypedImp(impSelf, Self.selector)
+                    }
+                },
+                onlyIfNonSwizzled: true
+            )
+        }
+    }
+}
+
+private extension URLRequest {
+    func merging(_ httpHeaders: [String: String]) -> URLRequest {
+        let mergedHeaders: [String: String]
+        if let currentHeaders = allHTTPHeaderFields {
```

Comment:
We should rather use the precise [`addValue(_:forHTTPHeaderField:)`](https://developer.apple.com/documentation/foundation/urlrequest/2011522-addvalue) from `URLRequest`. It does [internal synchronization](https://github.com/apple/swift-corelibs-foundation/blob/1c1e52abee4efc2e24def6ed117751909bbb32cb/Sources/FoundationNetworking/URLRequest.swift#L165-L173) of header fields dictionary, so it's thread safe. I think this is to not conflict with internal processing of request by `URLSession`.